### PR TITLE
various ports: remove subports/variants for EOL Python versions

### PR DESCRIPTION
**Pull request by reneeotten**
#### Description
This PR removes subports and/or variants for EOL Python versions. The (large) set of ports is internally consistent so that (hopefully) there are no ports left that have a dependency on a removed subport.

No other changes (except for removal of redundant fields) so if a port doesn't build anymore it is (or at least should) not caused by this PR.

###### Type(s)

- [ ] bugfix
- [x] enhancement
- [ ] security fix

###### Tested on
macOS 10.15.7 19H1824 x86_64
Xcode 12.4 12D4e

###### Verification 
Have you

- [x] followed our [Commit Message Guidelines](https://trac.macports.org/wiki/CommitMessages)?
- [x] squashed and [minimized your commits](https://guide.macports.org/#project.github)?
- [x] checked that there aren't other open [pull requests](https://github.com/macports/macports-ports/pulls) for the same change?
- [ ] referenced existing tickets on [Trac](https://trac.macports.org/wiki/Tickets) with full URL? 
- [x] checked your Portfile with `port lint --nitpick`?
- [ ] tried existing tests with `sudo port test`?
- [ ] tried a full install with `sudo port -vst install`?
- [ ] tested basic functionality of all binary files?

### CODE REVIEW

**Comment by Schamschula** [comment]
I'm glad to see this! I hadn't deleted these subports myself, as they still had dependents.

**Comment by chrstphrchvz** [comment]
Notifying maintainers not notified by @macportsbot (https://trac.macports.org/ticket/63821):

@blair for port py-tensorflow_estimator1.
@dfm for port py-corner.
@smithsp for port py-fortranformat.
@mojca for port py-django-countries, py-paho-mqtt, py-pycson, py-pypeg2, py-sounddevice, py-spacepy, py-speg.
@zdw for port py-dockerpy-creds.
@essandess for port py-graph-tool.
@jswhit for port py-pygrib.
@SoapZA for port py-pysam.
@ierofant for port py-bitstruct, py-canmatrix.
@donkopotamus for port py-dispatcher.

**Comment by jjstickel** [comment]
LGTM! Thanks for doing this.

File: python/py-fortranformat/Portfile
Comment (smithsp) [inline python/py-fortranformat/Portfile:22]: I approve.

**Review by smithsp — APPROVED** [review]
I approve.

**Review by nerdling — APPROVED** [review]

**Comment by kurthindenburg** [comment]
If you've verified no other ports need them, go ahead.  I know a lot of mine had a lot of other ports using them.

**Comment by xeron** [comment]
There're some ports outside of python group which depend on these older version, they may need a cleanup too, e.g.:

https://github.com/macports/macports-ports/blob/d275a24edbefe7f361f2aa6eb8f59648092058c0/databases/pgcli/Portfile#L24-L25
https://github.com/macports/macports-ports/blob/d275a24edbefe7f361f2aa6eb8f59648092058c0/databases/mycli/Portfile#L24-L25

**Comment by cjones051073** [comment]
This looks good to me.

I also vote for merging quick and if deps from variants etc. then pop on address them later on. Merging this is the easiest way to find them ;)

**Review by xeron — APPROVED** [review]

**Review by cjones051073 — APPROVED** [review]

**Review by mazhe — APPROVED** [review]

**Review by mrdomino — APPROVED** [review]

**Comment by blair** [comment]
I approve (I don't see an approve button where).

**Review by bgilbert — APPROVED** [review]

**Review by catap — APPROVED** [review]

**Review by blair — APPROVED** [review]

File: python/py-construct/Portfile
Comment (michaelld) [inline python/py-construct/Portfile:34]: Should this be removed?

**Review by jandemter — APPROVED** [review]

**Review by michaelld — COMMENTED** [review]
I am ambivalent about removing support for Py27. I have no objection to removing Py36 & TBH even Py37 if we want to. What about `py-scipy` and `py-numpy` And also the port I comment on here with respect to Py27 removal?

**Comment by cjones051073** [comment]
> I approve (I don't see an approve button in the UI).

View the changes, then Start a review and just submit your approval there with no comments.

File: python/py-construct/Portfile
Comment (reneeotten) [reply python/py-construct/Portfile:34]: no, the py27 subport will need to stay at least for a while since it has a dependent:
```
% port echo depends:py27-construct
gr37-satellites
```

**Comment by reneeotten** [comment]
> I am ambivalent about removing support for Py27. I have no objection to removing Py36 & TBH even Py37 if we want to. What about `py-scipy` and `py-numpy` And also the port I comment on here with respect to Py27 removal?

in the long run, yes these should/will be removed - but one has to start with the ones that have no dependents and work towards the ones higher in the food-chain like `py-numpy` and `py-scipy`.

**Comment by mascguy** [comment]
Renee, you might want to use `[skip ci]` in the commit message, as the CI jobs are highly unlikely to complete successfully. (And the resources will also be tied up for six hours.)

**Review by kurthindenburg — APPROVED** [review]

**Comment by reneeotten** [comment]
> Renee, you might want to use `[skip ci]` in the commit message, as the CI jobs are highly unlikely to complete successfully. (And the resources will also be tied up for six hours.)

sure, but where? Adding it for one random commit in the whole list will suffice, or first, last, ....?

**Comment by mascguy** [comment]
> > Renee, you might want to use `[skip ci]` in the commit message, as the CI jobs are highly unlikely to complete successfully. (And the resources will also be tied up for six hours.)
> 
> sure, but where? Adding it for one random commit in the whole list will suffice, or first, last, ....?

Good point, might be better to simply cancel CI manually then!  :-)

**Review by mndavidoff — APPROVED** [review]

**Review by cardi — APPROVED** [review]

**Review by amake — APPROVED** [review]

**Review by mohd-akram — APPROVED** [review]

File: python/py-construct/Portfile
Comment (michaelld) [reply python/py-construct/Portfile:34]: Gotcha. Somehow I hadn't made the connection that you're removing Py27 ports with no dependencies ... makes sense now!

**Review by michaelld — APPROVED** [review]
LGTM & makes sense to remove Py27 ports with no dependencies!

**Comment by lbschenkel** [comment]
I am OK with whatever the consensus ends up being.

**Review by lbschenkel — APPROVED** [review]

**Comment by i0ntempest** [comment]
No objections here

**Comment by nerdling** [comment]
> @jmroot Josh, do you have a TCL script to do a full exhaustive search of dependent ports, including all variants? (It's a bit of a pain for dependents of Python ports, given the use of `${python.version}` and such within names. Making a simple grep less-than-conclusive...)

Would `grep -r --include=Portfile py.*psycopg2` suffice?

**Comment by reneeotten** [comment]
> Well, in the interest of not causing pain for our users and maintainers, I think we need to do more exhaustive research before merging this. Particularly given that this isn't time-critical.

Frankly, I have done as much due diligence on this as I am willing/able to do given the amount of time it takes. If maintainers specifically object to the removal of Python versions for "their" ports I will take them out. If not, I intend to merge this PR after the usual three days waiting period. If anyone wants to go ahead and search the tree for occurrences in variants for the ~270 ports in this PR and list them here, please feel free - the clock is ticking ;) No seriously, I am fine with waiting a bit longer to merge this if someone is committing to do that work - but given how these exercises happened in the past, I don't hold my breath. 

Of course, I am more than happy to fix any possible, minor fallout that this could cause *after* the fact; I tried to be careful and not break anything with this PR similarly as I've done in the past.

**Comment by mascguy** [comment]
> > Well, in the interest of not causing pain for our users and maintainers, I think we need to do more exhaustive research before merging this. Particularly given that this isn't time-critical.
> 
> Frankly, I have done as much due diligence on this as I am willing/able to do given the amount of time it takes.

Well, that's why I added @jmroot to the discussion: There might be a TCL script to handle this, and/or built-in functionality in base.

So you needn't spend more time on it Renee, as we're offering to help. Just don't merge until others have had a chance to help you proactively vet the dependencies!  ;-)

**Review by JacksonIsaac — APPROVED** [review]
lg2m for the ports I am in maintainer list

**Comment by reneeotten** [comment]
I have rebased with `master` and fixed the conflicts; not sure why the GitHub website still thinks that it cannot be rebased... 

I'll give this another day so that people had more than a week to respond and then will merge this PR provided that I can figure out how to resolve this, or commit directly if that works out-of-the-box.

**Comment by mascguy** [comment]
> @jmroot Josh, do you have a TCL script to do a full exhaustive search of dependent ports, including all variants? (It's a bit of a pain for dependents of Python ports, given the use of `${python.version}` and such within names. Making a simple grep less-than-conclusive...)

@jmroot, do you have a TCL script to run an exhaustive dependency search like this, across all variants?

File: python/py-emcee/Portfile
Comment (lpsinger) [inline python/py-emcee/Portfile:34]: Since we have dropped support for Python <= 3.6, the condition in the `if` statement is always true, so the `conflicts` statement needs to be kept.

File: python/py-gcn/Portfile
Comment (lpsinger) [inline python/py-gcn/Portfile:9]:
```suggestion
python.versions     37 38 39 310
```

Comment (lpsinger) [inline python/py-gcn/Portfile:9]:
```suggestion
python.versions     37 38 39 310
```

**Review by lpsinger — CHANGES_REQUESTED** [review]

File: python/py-emcee/Portfile
Comment (reneeotten) [reply python/py-emcee/Portfile:34]: good catch, I missed that one - fixed locally!

File: python/py-gcn/Portfile
Comment (reneeotten) [reply python/py-gcn/Portfile:9]: while I in principle agree with adding subports, I will not do so in this PR which is purely about removing EOL versions. Adding newer Python subports requires one to make sure upstream supports them and such, that's something for a maintainer to do at another time.

File: python/py-emcee/Portfile
Comment (lpsinger) [reply python/py-emcee/Portfile:34]: It's not showing up in GitHub yet...

Comment (reneeotten) [reply python/py-emcee/Portfile:34]: correct, as I said “fixed locally” and I haven’t pushed yet to this PR

**Comment by mascguy** [comment]
Rene, I'm good with merging as well. As you said before, it's easy enough to fix anything broken after-the-fact.

**Review by mascguy — APPROVED** [review]

**Comment by mascguy** [comment]
Renee, it looks like port `volatility` is now broken, due to the removal of `py27-distorm`:

```
$ port deps volatility
Full Name: volatility @2.5_1
Build Dependencies:   py27-setuptools
Library Dependencies: py27-pycryptodome, py27-distorm, python27
Runtime Dependencies: yara

$ port rdeps volatility
Error: Port py27-distorm not found
```

So either `volatility` needs to be updated to use a later version of `py-distorm`, or the Python 2.7 version needs to be restored.

**Comment by mascguy** [comment]
> So either `volatility` needs to be updated to use a later version of `py-distorm`, or the Python 2.7 version needs to be restored.

It looks like `py-distorm` was removed entirely, which I didn't expect. So that ports needs to be restored ASAP.

**Comment by nerdling** [comment]
Or we can get rid of `volatility`.

**Comment by mascguy** [comment]
> Or we can get rid of `volatility`.

While I can't speak to how widely it's used by our users, I do use it personally. So I'm going to veto that idea!  ;-)

**Comment by jmroot** [comment]
There seems to have been some confusion between `distnorm` and `distorm`. The commit restores the correct name, but it looks like the other changes should be kept?

**Comment by mascguy** [comment]
> There seems to have been some confusion between `distnorm` and `distorm`. The commit restores the correct name, but it looks like the other changes should be kept?

The correct name appears to be `distorm`, per the project site: https://github.com/gdabah/distorm

`distnorm` appears to be a typo. The portfiles are also essentially identical, so `distnorm` needs to be removed. And dependents, if any, need to be updated to reference `py-distorm`

**Comment by jmroot** [comment]
Yes, as I said, the commit restored the correct name but also reverted some other changes.

**Comment by mascguy** [comment]
> Yes, as I said, the commit restored the correct name but also reverted some other changes.

My commit only involved resurrection of `py-distorm`. What "other changes" are you referring to?

**Comment by mascguy** [comment]
> > > Yes, as I said, the commit restored the correct name but also reverted some other changes.
> > 
> > 
> > My commit only involved resurrection of `py-distorm`. What "other changes" are you referring to?
> 
> Looks to me like it did more
> 
> [20fc94e](https://github.com/macports/macports-ports/commit/20fc94e702fa3ba8fa88d9f7e8603525f5752725)
> 
> you added back the 35 and 36 supports. I've since removed them again.

Correct, because I restored the port, based on the state prior to Renee's changes. In particular, the inadvertent renaming of the port to `py-distnorm`.

**Comment by mascguy** [comment]
Renee, I don't have the time to review the commits in this PR, given the sheer number of changes.

Can you please audit this work, and verify that no other ports were inadvertently renamed?

**Comment by mascguy** [comment]
> > Correct, because I restored the port, based on the state prior to Renee's changes. In particular, the inadvertent renaming of the port to py-distnorm.
> 
> And we're saying only the accidental name change should have been reverted, not the other, intentional, changes.

Yes, that's well-understood, and noted.

However, my primary concern is ensuring we don't have a broken ports tree.

**Comment by jmroot** [comment]
> Can you please audit this work, and verify that no other ports were inadvertently renamed?

Doesn't look like it. But this is a good reminder of why changes that aren't strictly related should be avoided in mass commits like this.

**Comment by mascguy** [comment]
> > However, my primary concern is ensuring we don't have a broken ports tree.
> 
> Well, that has yet to be accomplished in the general sense, so good luck. :)

It's a labor of love, but I'm trying!  ;-)

**Comment by reneeotten** [comment]
oh well.... sorry about that! Just changing the `name` field to correct the typo though would have fixed all of this - no need for mass panic ;)

**Comment by mascguy** [comment]
> oh well.... sorry about that! Just changing the `name` field to correct the typo though would have fixed all of this - no need for mass panic ;)

Yeah, sorry for being a bit dramatic folks!  :-D

**Comment by adfernandes** [comment]
Thanks for this, @reneeotten!

**Pull request merged**
